common/remote/sockfile.cpp

```diff
@@ -3811,17 +3831,244 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
+                {
+                    iFileIO.setown(iFile->open(IFOread));
+                    if(iFileIO)
+                        rowcompressed = true;
+                }
+            }
+            else
+#endif
+                iFileIO.setown(iFile->open(IFOread));
+            if (!iFileIO)
+                throw MakeStringException(0, "Failed to open: '%s'", fileName);
+
+            inputstream.setown(createFileSerialStream(iFileIO, startPos));
+            prefetchBuffer.setStream(inputstream);
+            prefetcher.setown(helper->queryDiskRecordSize()->createDiskPrefetcher(nullptr, 0));
+
+            outBuilder = new MemoryBufferBuilder(resultBuffer, helper->queryOutputMeta()->getMinRecordSize());
+            chooseN = helper->getChooseNLimit();
+            limit = helper->getRowLimit();
+        }
+
+        opened = true;
+    }
+    void close()
+    {
+        // TBD
+    }
+    bool segMonitorsMatch(const void *row) { return true; }
+public:
+    CRemoteDiskReadActivity(IPropertyTree &actNode) : prefetchBuffer(nullptr)
+    {
+        fileName.set(actNode.queryProp("fileName"));
+        inputFieldInfoDeserializer.setown(createRtlFieldTypeDeserializer());
+        outputFieldInfoDeserializer.setown(createRtlFieldTypeDeserializer());
+
+        const RtlRecordTypeInfo *inputRecordTypeInfo = getTypeInfo(*inputFieldInfoDeserializer, actNode, "input"); // NB: inputFieldInfoDeserializer owns returned object
+        const RtlRecordTypeInfo *outputRecordTypeInfo = getTypeInfo(*outputFieldInfoDeserializer, actNode, "output"); // NB: outputFieldInfoDeserializer owns returned object
+
+        outMeta.setown(new CDynamicOutputMetaData(*outputRecordTypeInfo));
+
+        unsigned __int64 chooseN = actNode.getPropInt64("choosen", defaultFileStreamChooseN);
+        unsigned __int64 skipN = actNode.getPropInt64("skipN", defaultFileStreamSkipN);
+        unsigned __int64 rowLimit = actNode.getPropInt64("rowLimit", defaultFileStreamRowLimit);
+        Owned<IOutputMetaData> in = new CDynamicOutputMetaData(*inputRecordTypeInfo);
+        helper.setown(createDiskReadArg(fileName, in.getClear(), outMeta.getLink(), chooseN, skipN, rowLimit));
+    }
+    ~CRemoteDiskReadActivity()
+    {
+        if (outBuilder)
+            delete outBuilder;
+    }
+    virtual const void *nextRow(size32_t &rowSz) override
+    {
+        if (!opened) open();
+        if (needTransform)
+        {
+            while (!eofseen && ((chooseN == 0) || (processed < chooseN)))
+            {
+                while (!prefetchBuffer.eos())
+                {
+                    prefetcher->readAhead(prefetchBuffer);
+                    const byte * next = prefetchBuffer.queryRow();
+                    size32_t sizeRead = prefetchBuffer.queryRowSize();
+                    if (segMonitorsMatch(next))
+                        rowSz = helper->transform(*outBuilder, next);
```

Comment:
very minor: more efficient to assign to a local, and then assign to the reference at the point it returns.

Reply:
will add local var.